test/EventParser.hs

```diff
@@ -0,0 +1,136 @@
+    case res of
+        Left err -> fail $ show err
+        Right x -> do
+           -- print x >> print (x == EVENT_HEADER_BEGIN)
+            when (x /= EVENT_HEADER_BEGIN ) $ fail "Event header begin not found."
+            (res2, nxt2) <- Stream.parseBreak word32be nxt
+            case res2 of
+                Left err -> fail $ show err
+                Right x -> do
+                    --print x >> print (x == EVENT_HET_BEGIN)
+                    when (x /= EVENT_HET_BEGIN ) $ fail "Event Type list, begin not found."
+                    (res2, nxt2) <- Stream.parseBreak word32be nxt
+                    ehe <- getEventTypes nxt2
+                    (res3, eds) <- Stream.parseBreak word32be ehe
+                    case res3 of
+                        Right EVENT_HEADER_END -> do
+                            print "EVENT_HEADER_END" >> print (x == EVENT_HEADER_END)
+                            (res, nxt) <- Stream.parseBreak word32be eds
+                            case res of
+                                Left err -> fail $ show err
+                                Right EVENT_DATA_BEGIN -> do
+                                    print "EVENT_DATA_BEGIN" >> getEventData nxt
+                        Right _ -> fail "Event header end not found."
+
+                    return ()
+            return ()
+    where
+
+    getEventData st = do
+        (ev_num, eds1) <- Stream.parseBreak word16be st
+        (res4, eds2) <- Stream.parseBreak word64be eds1
+        skl <-
+            case ev_num of
+                Right num -> do
+                    mp <- readIORef kvRef
+                    return $ Map.lookup (fromIntegral num) mp
+        (res5, eds3) <-
+            case skl of
+                Just l -> do
+                    skip l eds2
+                Nothing -> fail "Event type unknown."
+        print $ "Event num = " ++ show ev_num
+        print $ "evTime = " ++ show res4
+        getEventData eds3
+
+    getEventDatas st = do
+        (res, nxt) <- Stream.parseBreak word32be st
+        case res of
+            Left err -> fail $ show err
+            Right EVENT_DATA_BEGIN -> do
+                print "EVENT_DATA_BEGIN"
+                (res3, eds1) <- Stream.parseBreak word16be nxt
+                (res4, eds2) <- Stream.parseBreak word64be eds1
+                (res5, eds3) <- skip 14 eds2
+                (res6, eds4) <- Stream.parseBreak word16be eds3
+                (res7, eds5) <- Stream.parseBreak word64be eds4
+                (res8, eds6) <- Stream.parseBreak word32be eds5
+                print res3 >> print res4 >> print res6 >> print res7 >> print res8
+                return ()
+
+            Right EVENT_DATA_END -> print "EVENT_DATA_END"
+    getEventTypes st = do
+            (res, nxt) <- Stream.parseBreak word32be st
+            case res of
+                Left err -> fail $ show err
+                Right EVENT_ET_BEGIN -> getEventType nxt >>= getEventTypes
```

Comment:
This is an explicit loop, like C code, we should use parser combinators like `many` and Alternative instead.